runtime/lua/vim/shared.lua

```diff
@@ -10,6 +10,21 @@
 ---@diagnostic disable-next-line: lowercase-global
 vim = vim or {}
 
+--- Assert that a value has the expected type.
+---
+--- @generic T
+--- @param v T
+--- @param expected string Expected type
+local function expecttype(v, expected)
+  local actual = type(v)
+
+  -- Use an if statement with error() instead of assert() to avoid calling string.format if we don't
+  -- need it.
```

Comment:
Reference: https://chrisfls.github.io/luajit-wiki/Numerical-Computing-Performance-Guide/

> * Do not intersperse expensive or uncompiled operations.
>   - E.g. avoid assert(type(x) == "number", "x is a "..mytype(x)")
>     *  The problem is not the assert() or the condition (basically free). The problem is the string concatenation, which has to be executed every time, even if the assertion never fails!

Reply:
Put the reference in the comment? (May linkrot though.)